File: daemon_controller_child.py
#!/usr/bin/env python3
import os
import sys
import time
import multiprocessing
import selectors

NUM_PROCESS = 5

def main():
    # fork to daemonize the script
    pid = os.fork()

    if(pid):
        print("Daemonizing...\n")
        sys.exit(0)
    else:
        # we collect all the pipes to list
        pipes = []

        # selector to figureout which pipes to read
        # and which pipes are ready to read
        selector = selectors.DefaultSelector()

        for proc in range(NUM_PROCESS):
            # create the pipes
            pipes.append(multiprocessing.Pipe())

            if not os.fork(): # we're inside the child process

                # close the read pipe, we don't need to read from the 
                pipes[proc][0].close()
                count = 0
                while True:
                    count += 1
                    # write to the pipes 
                    pipes[proc][1].send(
                        "Process {proc}: {count}".format(
                            proc=proc, 
                            count=count
                        )
                    )
                    time.sleep(proc)
            else:
                pipes[proc][1].close()
                selector.register(pipes[proc][0], selectors.EVENT_READ) 

                # once we spawn the last child process, start listening to 
                # pipes for data.
                if proc == NUM_PROCESS-1: 
                    while True:
                        events = selector.select(timeout=1)
                        for key, mask in events:
                            print(key.fileobj.recv())
                        time.sleep(2)

if __name__ == "__main__":
    main()

File: daemon_controller_child.pl
#!/usr/bin/perl -w

use IO::Handle;
use IO::Select;

# daemonize

if (! fork()) {
    $0 = "controller";

    $sel = IO::Select->new();
    # spawn five processes
    foreach((1..5)) {
        pipe(${"read$_"}, ${"write$_"});
        my $pid = fork();
        if (! $pid) {
            # close (STDERR);
            # close (STDOUT);

            $0 = "child_$_";

            close(${"read$_"});

            my $count = 0;
            while(1) {
                print(${"write$_"}, "from $_: $count\n");
                sleep(2);
                $count++;
            }
        } else {
            close(${"write$_"});
            $sel->add(${"read$_"}); 
            if ($_ == 5) {
                while(1) {
                    @readable = $sel->can_read(2);
                    foreach (@readable) {
                        print (<$_>);
                    }
                }
            }

        }
    }
} else {
    print ("Daemonizinng...");
    exit(0);
}

